### ISSUE TITLE
Async event subscriber not updating UI after async call

### ISSUE THREAD

**Issue by jamesthurley** [angular/angular#7381]
I have (essentially) the following code in a component of my app:
```
  public ngOnInit() {
    this.thing.changed.subscribe(() => this.update());
  }

  public async update() {
    try {
      this.a = 1;
      this.b = await this.dataService.getAsync();
      this.c = 2;
    }
    catch (error) {
      // blah
    }
  }
```

The TypeScript is being compiled to ES6 then run through Babel for ES5 (as TypeScript can't compile to ES5 with async/await yet).

The UI is displaying variables `a`, `b`, and `c`.  When the `changed` event is raised by the `thing` service, variable `a` is correctly displayed as `1`, but `b` and `c` do not update in the UI.  Console logging shows they are updated in the component.  If I then trigger an update manually (by clicking some unrelated button that updates the UI), `b` and `c` are rendered correctly.

I've put a breakpoint after the async call, and `Zone.run` is in the call stack.

I can try and create a minimal reproduction, but my first question is whether this behaviour is expected (and async event subscribers are simply not supported), or if this looks like a genuine bug.

**Comment by jamesthurley** [comment]
A further piece of information, if I add the code `setTimeout(() => {}, 1000);` within the `update` method but before the async call to `this.dataService.getAsync()`, the timeout triggers the UI to update and render all the values correctly.  If I move it to after the async call, the timeout has no effect.  

Indeed if I update a variable within the timeout (`setTimeout(() => {this.b = 4;}, 1000);`), the updated value is also not rendered if the `setTimeout` is called after the async call to `this.dataService.getAsync()`, but again works fine if it is before.  I imagine this is all reinforcing the same point, but thought it might help.

**Comment by jamesthurley** [comment]
I've tried to create a plunkr to reproduce this (using promises and continuations), but I haven't succeeded.  In this one you can click the button and it will update the first number immediately, and the second only after the promise completes after 5 seconds.  But the updates are being processed fine:

http://plnkr.co/edit/SvzCen4ePRzIxAsAN4Qq?p=preview

So there is something different in my app that is breaking Angular's UI updating.  I'll investigate more tomorrow.  Any pointers to what I should look for would be welcome.

**Comment by jpsfs** [comment]
Hi!

The issue is that the async call result is outside ngZone thus not triggering the UI update.
You can do something like:

```javascript
public constructor(private _ngZone: NgZone) {

}

public async update() {
    try {
      this.a = 1;
      let newBValue = await this.dataService.getAsync();

      this._ngZone.run(() => {
          this.b = newBValue;
          this.c = 2;
      });
    }
    catch (error) {
      // blah
    }
  }

```

**Comment by zoechi** [comment]
@jpsfs the behavior looks like the update runs outside the zone, the question is IMHO - why is this running outside the zone?

**Comment by jamesthurley** [comment]
@jpsfs, thanks, I made the change as you suggested and that does indeed fix the problem.  But as @zoechi points out, the question is why I've lost the zone.  I've done some more investigation, and to be honest I'm more baffled.

I replaced the call to `dataService.getAsync()` with a simple promise that resolved after 1 second with some data.  I removed the async/await code and just did `promise.then(r => this.b = r)`.  I get the same issue.  So it's nothing to do with async/await transpiling, and nothing to do with my data service.

Secondly, this problem isn't isolated to this once instance.. I've realised it is now all over my site. When I navigate to a page which loads and displays data, the data never displays.

This all worked a few days ago, so what changed?  My CI server failed the build because E2E tests were timing out. I couldn't reproduce so ended up wiping out my local repo and running through the CI server steps. Now I have the same issue locally and can see that the E2E test timeouts are caused by the UI not updating.  The example above is simply the first one I came across (and assumed it was isolated).

So something has messed up Zone, or Angular, and I'm struggling to understand what right now.  I've tried going back to earlier commits which passed on the CI server, but I see the same issue.  That is why I think it's an external dependency.

Perhaps someone familiar with the Zone or Angular code could give me some tips to debug this?

**Comment by jpsfs** [comment]
Hi 

@jamesthurley, @zoechi 
I understand your question but there are some limitations on Zone.js that are being addressed now by the Angular team.

A new version (0.6.x) was released a just a few hours ago and it will probably solve many of the problems you describe. https://github.com/angular/zone.js

Nevertheless, indeed some callbacks currently run outside angular zone. I don't know what is the lib you are using to make the async call (Angular's HTTP works fine) but jQuery AJAX calls, IndexedDb, ... all resolve outside the zone, leading to the behavior you describe.

Hopefully the behavior of Zone will stabilize and be more predictable.

Cheers,

**Comment by jamesthurley** [comment]
Hi @jpsfs, sorry, I didn't mean that question to sound like a complaint about Zone... I actually no longer think this is a zone issue exactly. I replaced my async call with:
```
      var promise = new Promise<any>((resolve, reject) => {
        setTimeout(() => {resolve({"some":"data"}); }, 1000);
      });
      promise.then((r) => {
        this.b = r;
      });
```

And it still failed to update the UI, so it's not the HTTP call that is the issue (which does use Angular's `Http`).  And indeed this has all worked solidly for many weeks, and only broke when I wiped out my local repo and re-cloned.. my angular/zone/rxjs dependencies all have fixed versions, so these can't have changed.  My current best theory is something else is interfering.  I'll keep investigating.

In the mean time, if you have any suggestions of where I could put some breakpoints... e.g. where Angular and Zone interact to trigger an update, that would be useful information.

Another tidbit.. I removed my `angular2-polyfills` import so it wouldn't run though Webpack, and referenced `https://code.angularjs.org/2.0.0-beta.7/angular2-polyfills.js` directly from a script tag, and that made the problem _worse_: just navigating between pages seems to have UI updating issues with that setup.  I don't know what that means other than I'm having a very bad day in the office.

**Comment by jamesthurley** [comment]
Resolved!  The culprit was Babel... I'm not sure what precisely, but I reverted all my Babel related npm modules back to a known good version and everything is functioning normally again.  Apologies that the title of this issue turned out to be nothing to do with the problem.

I reproduced the issue on the [AngularClass Angular2 Webpack Starter](https://github.com/AngularClass/angular2-webpack-starter) project so I can upload that minimal demonstration of the issue if it interests anyone.  I simply modified the project to update the UI using setTimeout as above, then modified the Webpack config to use Babel as part of the transpilation process... and suddenly setTimeout stops updating the page.

Thanks for the help @jpsfs and @zoechi.

**Comment by wizarrc** [comment]
@jamesthurley I ran into the same issue with babel. I figured this beast out finally! I will explain what is happening and what I was able to hack together to correct the issue.

I want to start out saying all the pieces I was using.

Windows 10 1607 Build 14393.51
AspNet Core 1.0 final
nodejs 6.3.1
npm 3.10.3

package.json

```
{
  "devDependencies": {
    "babel-loader": "^6.2.4",
    "babel-plugin-transform-runtime": "^6.12.0",
    "babel-polyfill": "^6.13.0",
    "babel-preset-es2015": "^6.13.2",
    "babel-preset-angular2": "^0.0.2",
    "babel-preset-stage-0": "^6.5.0",
    "ts-loader": "^0.8.2",
    "typescript": "^1.8.10",
    "webpack": "^1.13.1"
.
.
.
  },
  "dependencies": {
    "@angular/common": "^2.0.0-rc.5",
    "@angular/compiler": "^2.0.0-rc.5",
    "@angular/core": "^2.0.0-rc.5",
    "@angular/http": "2.0.0-rc.5",
    "aspnet-webpack": "^1.0.9",
    "babel-runtime": "^6.11.6",
    "rxjs": "^5.0.0-beta.6",
    "zone.js": "^0.6.12"
.
.
.
  }
}
```

.babelrc
```
{
  "plugins": [ "transform-runtime" ],
  "presets": [ "es2015", "stage-0", "angular2" ]
}
```

webpack.config
```
var path = require('path');
var webpack = require('webpack');
.
.
.

module.exports = {
    resolve: {
        extensions: ['', '.js', '.ts']
    },
    module: {
        loaders: [
            { test: /\.ts$/, include: /ClientApp/, loader: 'babel-loader!ts-loader' }
 .
.
.
        ]
    },
    ts: {
        compiler: 'typescript'
    }
    entry: {
        main: ['./ClientApp/boot-client.ts']
    },
    output: {
        path: path.join(__dirname, 'wwwroot', 'dist'),
        filename: '[name].js',
        publicPath: '/dist/'
    },
    plugins: [
        new webpack.DllReferencePlugin({
            context: __dirname,
            manifest: require('./wwwroot/dist/vendor-manifest.json')
        })
.
.
.
    ]
};
```

tsconfig.json
```
{
  "compilerOptions": {
    "moduleResolution": "node",
    "target": "es6",
    "sourceMap": false,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "skipDefaultLibCheck": true
  },
  "exclude": [
      "node_modules"
  ]
}
```

./wwwroot/index.cshtml
```
.
.
.
<app>Loading...</app>
<script src="~/dist/vendor.js" asp-append-version="true"></script>
@section scripts {
    <script src="~/dist/main.js" asp-append-version="true"></script>
}
```

./ClientApp/boot-client.ts
```
require('zone.js');
import 'bootstrap';
import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
import { AppModule } from './app.module;
.
.
.
platformBrowserDynamic().bootstrapModule(AppModule);
```

I cut out the unrelated parts of the file.

My understanding of zones.js is that it monkey patches all the various asynchronous calls in order to have angular 2 get notified and perform a change detection. So it turns out that if you import/require zones.js inside the boot-client.ts file while compiling async/await down to es5 using babel with the async plugin, in this case stage-0, it will wrap the async call in a promise. The problem is that babel injects code to patch the native Promise API before the zone.js has a chance to. A workaround is to patch the Promise using zone.js before the babel runtime has a chance to do so, then create a global variable of the patched zone-aware Promise API. Once inside the webpack bootstrap file, in this case, boot-client.ts, patch the babel runtime to use the zone-aware promise. I provide an example patch to solve this issue allowing babel to use a zone-aware promise. In the example I copy zone.js outside of the npm package and into the wwwroot/dist folder. Unfortunately, I can no longer update zone.js using npm. **Ideally I would like to keep zone.js inside webpack but outside of babel, but I'm not really sure how to achieve this.**

**Workaournd:**

index.cshtml
```
.
.
.
<app>Loading...</app>
<script src="~/dist/vendor.js" asp-append-version="true"></script>
@section scripts {
    <script src="~/dist/zone.min.js"></script>
    <script>
            _zonePromise = Promise;
    </script>
    <script src="~/dist/main.js" asp-append-version="true"></script>
}
```

./ClientApp/boot-client.ts
```
declare var _zonePromise;
require('babel-runtime/core-js/promise').default = _zonePromise;
import 'bootstrap';
import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
import { AppModule } from './app.module;
.
.
.
platformBrowserDynamic().bootstrapModule(AppModule);
```

Hope this hack helps anyone trying to use Angular2/TypeScript targeting es6 wanting async/await with babel and webpack.

**Comment by wizarrc** [comment]
@jpsfs Do you know if there is any good way to patch in the correct order without moving zone.js outside of webpack? The problem wasn't with angular, zone, nor with babel, but that both babel and zone both try to patch the promise object but in the wrong order. I can imagine the same issue that is happening with babel will happen with TypeScript 2.1 async -> es5 support in the coming months unless the TypeScript's non-runtime version will not patch promise in order to work correctly. Better project level support for all these libraries are needed.

**Comment by jamesthurley** [comment]
@wizarrc, amazing work! This problem re-appeared for me when I updated to Angular 2 Beta 8, even with the fixed Babel version, so since then I've had both Babel and Angular fixed to old versions.

My plan was to wait until TypeScript 2.0 (now bumped to 2.1) before attempting to update everything, sans Babel, and it sounds like I would have been disappointed with the result!  I'll try and give your fix a try in the near future.  I really appreciate you digging into this.. I'm surprised it hasn't effected more people.

**Comment by wizarrc** [comment]
@jamesthurley I have a nice workaround update. I found a way to keep zone.js inside the npm/webpack package management. Add a separate loader to your webpack.config.js file and raw load in the javascript workaround.


webpack.config.js
```
module: {
        loaders: [
            { test: /\.js$/, include: /ClientApp\/fixes/, loader: 'raw-loader' },
            { test: /\.ts$/, include: /ClientApp/, loader: 'babel-loader!ts-loader' },
.
.
.
        ]
    },
```

./ClientApp/fixes/zone-babel-async-fix.js
```
require("zone.js");

define(["require", "exports"], function (require, exports) {
    "use strict";
    require('babel-runtime/core-js/promise').default = Promise;
    exports.zonePromise = Promise
});
```

./ClientApp/boot-client.ts
```
require('./fixes/zone-babel-async-fix');
import 'reflect-metadata';
import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
import { AppModule } from './app.module';

platformBrowserDynamic().bootstrapModule(AppModule);
```

**Comment by wizarrc** [comment]
@jamesthurley I updated the previous post, I forgot to paste in the entry update in webpack.config.js

**Comment by wizarrc** [comment]
@jamesthurley Working with the configuration provided, I came across another issue that seems to be a hit or miss with the babel-loader I can't quite narrow down. babel generates bad code (seems to happen when I refactor my code a lot). It generates "import _Object$defineProperty from 'babel-runtime/core-js/object/define-property';" that is still es6. I'm guessing this is still not quite stable.

Looking around I saw that others have hit the same issue and have had a workaround provided to isolate the plugins. I've changed my .babelrc configure to incorporate the workaround. Let me know if you hit this error yourself.

https://github.com/babel/babel-loader/issues/195#issuecomment-238804184

.babelrc
```
{
  "passPerPreset": true,
  "presets": [
    { "plugins": [ "transform-runtime" ] },
    {
      "passPerPreset": false,
      "presets": [ "es2015", "stage-0", "angular2" ]
    }
  ]
}
```

**Comment by jamesthurley** [comment]
Hi @wizarrc, I tried this fix today and got it to work which is a huge help to me, so thanks again for sharing the solution.

I had to tweak the solution slightly.  I was trying your second version of the fix which allowed keeping `zone.js` in npm, but it wasn't having any effect. To verify things I stuck a `console.log` in `zone-babel-async.js`, and determined it wasn't being executed at all. However from my understanding of `raw-loader` this would make sense, as `raw-loader` would simply return the contents of the js file as a string, right? But if it's working for you I might be wrong there!

So instead of doing `require('./fixes/zone-babel-async-fix');` and using `raw-loader`, I added another entry point in `webpack.config.json`:

    entry: {
      'fixes': './src/fixes/zone-babel-async.js',
      'vendor': './src/vendor.ts',
      'app': './src/bootstrap.ts'
    }

My webpack config is set up not to use Babel for `.js` files anyway, so this allowed the fix to run without going through Babel first.  Everything is now running great again.

I haven't hit the other error you mentioned yet...

**Comment by wizarrc** [comment]
@jamesthurley https://github.com/angular/angular/issues/7381#issuecomment-240753388

You are absolutely correct. As you can see, I forgot to update that portion of the config till about 10 minutes after I posted. Next time try hitting refresh ;) The only difference is that I grouped mine in an array instead of tagging each one separately.

Glad to see it's working for you!

**Comment by jamesthurley** [comment]
Ha, you're right! I'm so used to pages updating themselves these days I forgot to refresh the page manually :)

**Comment by wizarrc** [comment]
@jamesthurley Does your output now include 3 separate output files in your "wwwroot/dist" folder instead of just 2 (in your case app.js, vendor.js, and now fixes.js). My solution merges the app.js and lets webpack bring it in with a require. Each way works just as well, depends on how you prefer to separate your output.

**Comment by jamesthurley** [comment]
@wizarrc, yep you're right, I had ended up with three separate files.  I've now merged them back into two as you suggested.

**Comment by Luchillo** [comment]
Hi, i'm having a similar issue, i'm using a function that triggers an `NgClass` expression, that function is used in between 2 `async/await` functions and it updates in code but UI doesn't update.

I'm using webpack with Awesome Typescript Loader (just in case because i don't use babel).

Here's how it looks:

```ts
    let printerDevice: string = await this.conf.showRadioButtons({
        title: `Seleccione dispositivo usb`,
        inputs: [
          // { name: 'dataTable', label: 'dataTable', type: 'text', placeholder: '13' },
          ...map(devices, (device) => ({ label: device, value: device, type: 'radio' }))
        ]
      });

      if (!printerDevice) {
        throw new ReferenceError('undefined device');
      }

      this.conf.showProgress(); // <- Line that tries to update the UI

      await this.printerService.connect(printerDevice);
```

So i was thinking either `async/await` halts UI js context execution, or it has the original issue where it runns out of the angular 2 zone.

**Comment by Zaniyar** [comment]
So by last quarter of 2018 we still have the same issue with async calls and zone.js.

**Comment by wizarrc** [comment]
@Zaniyar I'd imagine the issue isn't inherently fixed as it's really doing the correct thing. zone.js by design needs to patched outside of a babel, so the workaround is the correct thing to do when using babel. As later Angular was updated to support later versions of TypeScript, the issue is now moot as TypeScript now transpiles async/await to ES5 without the need for babel.

**Comment by bill-sama9008** [comment]
I have the same problem with the version of angular 6.1.10. It does not update the form data. It is only updated if I click on some field of the form which should not happen. This happens to me when I make a guard in a component that validates the user's authentication status and I subscribe to a firebase route, but if I remove the guard it works correctly for me. I have been looking at all the logic and so far I have not been able to solve the problem

**Comment by guftall** [comment]
I had same problem with angular 6 and 7. There was a `*ngFor` in my html file and when i update the array, nothing happened to ui until click on one of components in that page. so **this worked for** me:

1. instead of simple `*ngFor` on an array, use an `Observable` and `let a of _observableVar | async` in html file
2. inject a `private _ngZone: NgZone` in `component` constructor
3. and write this line `this._ngZone.run(() => {});` after your async call.

neither of my methods have `async/await` syntax


```javascript
  getListFromServer() {
    this._observableVar = this.api.listAvailableAndReservedTurns();
    this._ngZone.run(() => {});
  }
```
